src/inmanta/deploy/timers.py

```diff
@@ -0,0 +1,295 @@
+        """
+
+        # Create if it is not known yet
+        if resource not in self.resource_timers:
+            self.resource_timers[resource] = ResourceTimer(resource, self._resource_scheduler)
+
+        def _setup_repair(repair_interval: int) -> tuple[int, str, TaskPriority]:
+            countdown = repair_interval
+            reason = (
+                f"Individual repair triggered for resource {resource} because last "
+                f"repair happened more than {countdown}s ago."
+            )
+            priority = TaskPriority.INTERVAL_REPAIR
+
+            return (countdown, reason, priority)
+
+        def _setup_deploy(deploy_interval: int) -> tuple[int, str, TaskPriority]:
+            countdown = deploy_interval
+            reason = (
+                f"Individual deploy triggered for resource {resource} because last "
+                f"deploy happened more than {countdown}s ago."
+            )
+            priority = TaskPriority.INTERVAL_DEPLOY
+
+            return (countdown, reason, priority)
+
+        # Both periodic repairs and deploys are disabled on a per-resource basis.
+        if self.periodic_repair_interval is None and self.periodic_deploy_interval is None:
+            return
+
+        if not is_dirty:
+            # At the time of the call, the resource is in an assumed good state:
+            # schedule a periodic repair for it if the repair setting is set on a per-resource basis
+            if self.periodic_repair_interval:
+                countdown, reason, priority = _setup_repair(self.periodic_repair_interval)
+            else:
+                return
+
+        else:
+            # At the time of the call, the resource is in an assumed bad state:
+            # schedule a repair or a deploy, whichever has the shortest interval set
+            # on a per-resource basis.
+
+            if self.periodic_repair_interval and self.periodic_deploy_interval:
+                if self.periodic_repair_interval < self.periodic_deploy_interval:
+                    countdown, reason, priority = _setup_repair(self.periodic_repair_interval)
+                else:
+                    countdown, reason, priority = _setup_deploy(self.periodic_deploy_interval)
+            elif self.periodic_repair_interval:
+                countdown, reason, priority = _setup_repair(self.periodic_repair_interval)
+            else:
+                assert self.periodic_deploy_interval is not None  # mypy
+                countdown, reason, priority = _setup_deploy(self.periodic_deploy_interval)
+
+        self.resource_timers[resource].cancel()
+        self.resource_timers[resource].set_timer(
+            countdown=countdown,
+            reason=reason,
+            priority=priority,
+        )
+
+    def remove_resource(self, resource: ResourceIdStr) -> None:
+        """
+        Stop managing the timer for the given resource and
+        cancel the associated re-deployment, if any.
```

Comment:
Didn't you propose on Slack that the timer instance would be kept in existence, only cancelling its timer? I'm fine either way tbh, this deviation just surprised me.

Reply:
Right, I think I found this way easier to reason about at some point and then it stuck around ? But now I wonder about performance tradeoff between both approaches.

I think deleting+recreating has more overhead but then all operations on the resource_timers dict would be faster since the set is smaller ?

Reply:
> all operations on the resource_timers dict would be faster since the set is smaller ?

I think that direct operations should have very little performance difference (could be wrong). Operations that only act on the *running* timers are a different matter, but we could easily keep a set of those.

That said, I think it would be OK to leave it as is, I'm not sure that the performance overhead would actually be significant. But I'd quickly run it by @wouterdb since it does deviate from what was concluded between the two of you on Slack.